lib/chef/provider/package.rb

```diff
@@ -438,45 +438,47 @@ def target_version_array
         @target_version_array ||=
           begin
             target_version_array = []
-            each_package do |package_name, new_version, current_version, candidate_version|
+            each_package do |package_name, new_version, current_version, candidate_version, magic_version|
               case action
               when :upgrade
-                if current_version.nil?
-                  # with use_magic_version there may be a package installed, but it fails the user's
-                  # requested new_resource.version constraints
+                if version_equals?(current_version, new_version)
```

Comment:
Can you add a comment to each one of these conditionals on why it is in the place it is in. Stuff like not un-lazying `candidate_version`, etc., specifically is worth mentioning. Why to do magic_version before other things, etc.